other/zombo-bakery/src/lib.rs

```diff
@@ -0,0 +1,228 @@
+
+            // Return the NFT, change, and zombo
+            (nft_bucket, payment, zombo_bucket)
+        }
+        
+        //Need NFT to purchase flour
+        #[auth(baker_nft_def)]
+        //Buy some organic whole wheat flour
+        pub fn buy_flour(&self, payment:Bucket) -> (Bucket,Bucket) {
+            //take payment in zombo
+            self.zombo.put(payment.take(self.material_price));
+            info!("{}", "Enjoy the flour!");
+            //return flour and remaining zombo
+            (self.material_minter.authorize(|auth| self.flour.mint(1, auth)),payment)
+        }
+
+        //Need NFT to purchase water
+        #[auth(baker_nft_def)]
+        //Buy some distilled water
+        pub fn buy_water(&self, payment:Bucket) -> (Bucket,Bucket) {
+            //take payment in zombo
+            self.zombo.put(payment.take(self.material_price));
+            info!("{}", "Enjoy the water!");
+            //return water and remaining zombo
+            (self.material_minter.authorize(|auth| self.water.mint(1, auth)),payment)
+        }
+
+        //Need NFT to purchase yeast
+        #[auth(baker_nft_def)]
+        //Buy some Fleischmann's bakers yeast
+        pub fn buy_yeast(&self, payment:Bucket) -> (Bucket, Bucket) {
+            //take payment in zombo
+            self.zombo.put(payment.take(self.material_price));
+            info!("{}", "Enjoy the yeast!");
+            //return water and remaining zombo
+            (self.material_minter.authorize(|auth| self.yeast.mint(1, auth)),payment)
+        }
+
+        //Combine all ingredients in bread maker
+        //Need flour water and yest to make bread
+        pub fn make_bread(&self, flour:Bucket, water:Bucket, yeast:Bucket) -> Bucket {
+            assert!(water.resource_def() == self.water, "That aint flour!");
+            assert!(flour.resource_def() == self.flour, "That aint water!");
+            assert!(yeast.resource_def() == self.yeast, "That aint yeast!");
+
+            //burn flour water and yeast and mint bread
+            self.material_minter.authorize(|auth| self.water.burn_with_auth(water, auth));
+            self.material_minter.authorize(|auth| self.flour.burn_with_auth(flour, auth));
+            self.material_minter.authorize(|auth| self.yeast.burn_with_auth(yeast, auth));
+            info!("{}", "Enjoy your bread, careful it's hot!!!");
+            self.material_minter.authorize(|auth| self.bread.mint(1, auth))
+        }
+
+        //Breads always better with butter
+        //Need bread to purchase butter
+        #[auth(bread)]
+        pub fn buy_butter(&self,payment:Bucket)->(Bucket,Bucket){
+            info!("{}", "Bread and butter is the best!");
+            (self.material_minter.authorize(|auth| self.butter.mint(1, auth)),payment)
+        }
+
+        //Enjoy your bread
+        //Need bread butter and NFT to eat bread
+        pub fn eat_bread(&self, bread:Bucket, butter:Bucket, nft_bucket:Bucket)->Bucket{
```

Comment:
For the `nft_bucket`, you could request a BucketRef instead of a Bucket. That way, you assure the user that you are not going to steal their badge and store it in your vault. Also, you will not have to return the bucket back. You only have to drop it at the end: `nft_bucket.drop()`